/* Copyright 2023 The FedLearner Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto3";
import "fedlearner_webconsole/proto/common/extension.proto";
package fedlearner_webconsole.proto;

// Cas more detail in
// https://apereo.github.io/cas/5.1.x/protocol/CAS-Protocol.html
message CasProtocol {
  string cas_server_url = 1;
  string service_url = 2;
  string login_route = 3;
  string validate_route = 4;
}

message OAuthProtocol {
  // Client id of our app
  string client_id = 1;
  // URLs of OAuth server
  string authorize_url = 2;
  string access_token_url = 3;
  string user_info_url = 4;
  string logout_url = 5;
  string redirect_uri = 6;
  // Url for FE getting code
  string code_key = 7;
  // Not visible for frontend
  string secret = 8 [(fedlearner_webconsole.proto.secret) = true];
  string username_key = 9 [(fedlearner_webconsole.proto.secret) = true];
  string email_key = 10 [(fedlearner_webconsole.proto.secret) = true];
}

message CustomProtocol {
  // Not visible for frontend
  string access_key = 1;
  string authorize_url = 2;
}

message Sso {
  string name = 1;
  string icon_url = 2;
  string protocol_type = 3;
  oneof protocol {
    CasProtocol cas = 4;
    OAuthProtocol oauth = 5;
    CustomProtocol custom = 6;
  }
  string display_name = 7;
}

message User {
  int64 id = 1;
  string username = 2;
  string password = 3 [(fedlearner_webconsole.proto.secret) = true];
  // Name of the role enum
  string role = 4;
  string name = 5;
  string email = 6;
  // Name of the state enum
  string state = 7;
  string sso_name = 8;
  int64 last_sign_in_at = 9;
  int64 failed_sign_in_attempts = 10;
}

message SigninParameter {
  string username = 1;
  string password = 2;
  string code = 3;
  string ticket = 4;
}
